File: PythonAlgorithmInterView/문자열조작/가장긴팰린드롬.js
/**
 * @param {string} s
 * @return {string}
 * 5번 문제
 */
 var longestPalindrome = function(s) {

    const lenPalindrome = (left, right) => {

        while(left >= 0 && right <= s.length && s[left] == s[right]){
            left -= 1;
            right += 1;
        }

        return s.slice(left+1, right)

    }


    if( s.length < 2 || s ==  s.split('').reverse().join('')){
        return s;
    }

    result = ''

    for(let i = 0; i < s.length; i++){
        let arr = [result, lenPalindrome(i, i+1), lenPalindrome(i, i+2)]
        arr.sort((a,b) => b.length - a.length);
        result = arr[0]
    }

    return result

};

let s = 
'babad'

console.log(longestPalindrome(s))

File: PythonAlgorithmInterView/문자열조작/가장긴팰린드롬.py


def longestPalindrome(self, s: str) -> str:

    def lenPalindrome(left, right) :
        print(left, right)
        while left >= 0 and right <= len(s) -1 and s[left] == s[right]:
            left -= 1
            right += 1
        return s[left + 1 : right]

    if len(s) < 2 or s == s[::-1]:
        return s

    result = ''
    for i in range(len(s)):
        result = max(
            result,
            lenPalindrome(i, i+1),
            lenPalindrome(i, i+2),
            key=len
        )
    return result
if __name__ == '__main__':
    s = "babad"
    print(longestPalindrome(None, s))